notebooks/Numba_051_release_demo.ipynb

```diff
@@ -0,0 +1,564 @@
+    "    @njit(cache=True)\n",
+    "    def specialised(x):\n",
+    "        return specialise_on_this_function(x)\n",
+    "    return specialised\n",
+    "\n",
+    "@njit(cache=True)\n",
+    "def f(x):\n",
+    "    print(\"f(x)\", x)\n",
+    "\n",
+    "special_f = make_function(f)\n",
+    "special_f(10)\n",
+    "special_f.stats # cache miss as it had to compile it, but no complaints about doing the caching!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# User defined mutable pass-by-ref structures\n",
+    "\n",
+    "A very common question from users is:\n",
+    "\n",
+    "> What can I use as a mutable structure that's also pass-by-reference?\n",
+    "\n",
+    "the answer is the new ``StructRef`` type (warning: this is experimental!), [documentation is here](http://numba.pydata.org/numba-doc/dev/extending/high-level.html#implementing-mutable-structures)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from numba.experimental import structref\n",
+    "\n",
+    "\n",
+    "# Define a StructRef.\n",
+    "# `structref.register` associates the type with the default data model.\n",
+    "# This will also install getters and setters to the fields of\n",
+    "# the StructRef.\n",
+    "@structref.register\n",
+    "class FruitType(types.StructRef):\n",
+    "    def preprocess_fields(self, fields):\n",
+    "        # This method is called by the type constructor for additional\n",
+    "        # preprocessing on the fields.\n",
+    "        # Here, we don't want the struct to take Literal types.\n",
+    "        return tuple((name, types.unliteral(typ)) for name, typ in fields)\n",
+    "\n",
+    "\n",
+    "# Define a Python type that can be used as a proxy to the StructRef\n",
+    "# allocated inside Numba. Users can construct the StructRef via\n",
+    "# the constructor for this type in python code and jit-code.\n",
+    "class Fruit(structref.StructRefProxy):\n",
+    "    def __new__(cls, kind, amount):\n",
+    "        # Overriding the __new__ method is optional, doing so\n",
+    "        # allows Python code to use keyword arguments,\n",
+    "        # or add other customized behavior.\n",
+    "        # The default __new__ takes `*args`.\n",
+    "        # IMPORTANT: Users should not override __init__.\n",
+    "        return structref.StructRefProxy.__new__(cls, kind, amount)\n",
+    "\n",
+    "    # By default, the proxy type does not reflect the attributes or\n",
+    "    # methods to the Python side. It is up to users to define\n",
+    "    # these. (This may be automated in the future.)\n",
+    "\n",
+    "    @property\n",
+    "    def kind(self):\n",
+    "        # To access a field, we can define a function that simply\n",
+    "        # return the field in jit-code.\n",
+    "        # The definition is shown later.\n",
+    "        return Fruit_get_kind(self)\n",
```

Comment:
writing this as the usual `return self.kind` crashes Python. Is this a bug? If not, a note highlighting that it must be done via an `njit` function would be helpful.

Reply:
The function is required to act as an accessor to the instance of the representation in the compiled code.